File: src/lib/api/parser.ts
import { Readable } from "stream";

export function parseVisura(testo: string): { ragione_sociale: string; ateco: string } {
  let ragione_sociale = "";
  let ateco = "";

  const rsMatch = testo.match(
    /(?:RAGIONE\s*SOCIALE|DENOMINAZIONE|IMPRESA)\s*:?\s*(.+?)(?:\n|$)/i
  );
  if (rsMatch) {
    ragione_sociale = rsMatch[1].trim().replace(/\s+/g, " ");
  }

  const atecoMatch = testo.match(
    /(?:CODICE\s*)?ATECO\s*:?\s*(\d{2}\.\d{2})/i
  );
  if (atecoMatch) {
    ateco = atecoMatch[1].trim();
  }

  return { ragione_sociale, ateco };
}

export async function extractPdfText(buffer: Buffer): Promise<string> {
  const pdfjsLib = await import("pdfjs-dist/legacy/build/pdf.mjs");
  const data = new Uint8Array(buffer);
  const doc = await pdfjsLib.getDocument({ data }).promise;
  const pages: string[] = [];
  for (let i = 1; i <= doc.numPages; i++) {
    const page = await doc.getPage(i);
    const content = await page.getTextContent();
    const pageText = content.items.map((item: any) => item.str).join(" ");
    pages.push(pageText);
  }
  await doc.destroy();
  return pages.join("\n");
}

export async function fileToBuffer(file: File): Promise<Buffer> {
  const arrayBuffer = await file.arrayBuffer();
  return Buffer.from(arrayBuffer);
}

export async function parseUploadedPdf(file: File): Promise<string> {
  const buffer = await fileToBuffer(file);
  return extractPdfText(buffer);
}

File: backend/services/parser.py
from io import BytesIO
from pypdf import PdfReader


def extract_pdf_text(content: bytes) -> str:
    reader = PdfReader(BytesIO(content))
    pages: list[str] = []
    for page in reader.pages:
        text = page.extract_text()
        if text:
            pages.append(text)
    return "\n".join(pages)


def parse_visura(testo: str) -> dict:
    import re
    ragione_sociale = ""
    ateco = ""

    rs_match = re.search(
        r"(?:RAGIONE\s*SOCIALE|DENOMINAZIONE|IMPRESA)\s*:?\s*(.+?)(?:\n|$)",
        testo, re.IGNORECASE
    )
    if rs_match:
        ragione_sociale = rs_match.group(1).strip()

    ateco_match = re.search(
        r"(?:CODICE\s*)?ATECO\s*:?\s*(\d{2}\.\d{2})",
        testo, re.IGNORECASE
    )
    if ateco_match:
        ateco = ateco_match.group(1).strip()

    return {"ragione_sociale": ragione_sociale, "ateco": ateco}
